provider/pkg/clients/fake/discovery.go

```diff
@@ -0,0 +1,191 @@
+// Copyright 2016-2024, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package fake
+
+import (
+	"embed"
+	_ "embed" // for embedded metadata documents
+	"encoding/json"
+	"io"
+	"io/fs"
+
+	openapi_v2 "github.com/google/gnostic-models/openapiv2"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	kubeversion "k8s.io/apimachinery/pkg/version"
+	"k8s.io/client-go/discovery"
+	discoveryfake "k8s.io/client-go/discovery/fake"
+	"k8s.io/client-go/restmapper"
+	kubetesting "k8s.io/client-go/testing"
+)
+
+//go:embed swagger.json
+var swagger []byte
+
+func LoadOpenAPISchema() (*openapi_v2.Document, error) {
+	return openapi_v2.ParseDocument(swagger)
+}
+
+//go:embed serverresources
+var serverresources embed.FS
+
+func loadServerResources() ([]*metav1.APIResourceList, error) {
+	all := []*metav1.APIResourceList{}
+	err := fs.WalkDir(serverresources, ".", func(path string, d fs.DirEntry, err error) error {
+		if !d.IsDir() && d.Name() == "serverresources.json" {
```

Comment:
nit: Could we invert this into a guard clause instead to reduce nesting. eg.
```suggestion
		if d.IsDir() || d.Name() != "serverresources.json" {
			return nil
		}
		<json unmarshaling logic>
```